clab/docker.go

```diff
@@ -179,6 +187,52 @@ func (c *cLab) CreateContainer(ctx context.Context, node *Node) (err error) {
 	return createContainerNS(node.Pid, node.LongName)
 }
 
+func (c *cLab) PullImageIfRequired(ctx context.Context, node *Node) (err error) {
+	filter := filters.NewArgs()
+	filter.Add("reference", node.Image)
+
+	ilo := types.ImageListOptions{
+		All:     false,
+		Filters: filter,
+	}
+
+	log.Debugf("Looking up %s Docker image", node.Image)
+
+	images, err := c.DockerClient.ImageList(ctx, ilo)
+	if err != nil {
+		return err
+	}
+
+	// Image doesn't exist, need to pull it
+	if len(images) == 0 {
+		// might need canonical name e.g.
+		//    -> alpine == docker.io/library/alpine
+		//    -> foo/bar == docker.io/foo/bar
+		//    -> docker.elastic.co/elasticsearch/elasticsearch == docker.elastic.co/elasticsearch/elasticsearch
+		canonicalImageName := node.Image
+		slashCount := strings.Count(node.Image, "/")
+
+		switch slashCount {
+		case 0:
+			canonicalImageName = "docker.io/library/" + node.Image
+		case 1:
+			canonicalImageName = "docker.io/" + node.Image
+		}
```

Comment:
when I looked into SDK examples I saw that they don't use the canonical name https://docs.docker.com/engine/api/sdk/examples/#pull-an-image

can you test if that is needed? My assumption is that if you specify simply `alpine` it will still do the pull

Reply:
I did and I was shocked myself.

```
Start - alpine
repository name must be canonical
Done - alpine

Start - library/alpine
repository name must be canonical
Done - library/alpine

Start - docker.io/library/alpine
{"status":"Pulling from library/alpine","id":"latest"}
...
{"status":"Pull complete","progressDetail":{},"id":"df20fa9351a1"}
Done - docker.io/library/alpine
```

Reply:
Interesting, I see that we use https://github.com/srl-wim/container-lab/blob/f11586e6916cb0bf0600633073e1bb001a6f4e4b/clab/clab.go#L8 as a docker api package, although this is a fork from `moby/moby`
the official Go docs actually say to use the `github.com/docker/docker/client` package

I wonder, can you swap it and give it a try? Maybe the upstream Go SDK is a more polished version that we should rather use?

/cc @karimra

Reply:
This was changed because of this issue #23

Reply:
yes, I remember #23, it just maybe it was not the right pointer =)

Reply:
The official `github.com/docker/docker/client` works fine as well and doesn't break anything, but the canonical name logic is still required.